### ISSUE TITLE
Thread safety

### ISSUE THREAD

**Issue by AnthonyLloyd** [fscheck/FsCheck#413]
Looking at the source it seems FsCheck is not thread safe wrt to use of  `Arb.arbitrary.Value`.

Is it possible to change this? I'm not sure if later use of Arb/Gen make this difficult.

I can't see any way of running tests in parallel without this change.

**Comment by kurtschelfthout** [comment]
What in particular is not thread safe?

**Comment by AnthonyLloyd** [comment]
Sorry, just realizing myself what is going on here. It relates to an issue (haf/expecto#123) in Expecto.

```fsharp
let internal arbitrary = new ThreadLocal<TypeClass<Arbitrary<obj>>>(fun () -> defaultArbitrary)
```

The above code means the tests are run in a thread safe way when they alter arbitrary.
`Arb.register<_>()` though makes the API essentially single threaded.

I can work around this by trying to replace this function but it won't be perfect since the order of the opens will be important. Or I could just recommend that `Arb.register<_>()` is not used because it doesn't support multi threading.

**Comment by kurtschelfthout** [comment]
`Arb.register` already merges the given value with the current `ThreadLocal` instance of the thread it's called in, which as a base value contains the default `Arbitrary` instances included in FsCheck. 

As long as you don't use async or `Task` which thread hops (which could lead to confusing results since registration in one task can affect another task) you should be fine. Even if you do use Task etc you can still use `Arb.register` as long as you make sure it's called before any tasks are run on the thread and not changed thereafter (i.e. you override instances once and for all your tests). 

Potentially we could improve that async situation by using `AsyncLocal` but I'm not sure it's honored by F#'s `async`.

**Comment by AnthonyLloyd** [comment]
Do you see `Arb.register` to be used just in the thread a test is going to run in and not a global register?

If I want to globally register a type and run tests in parallel how would I do that?

**Comment by kurtschelfthout** [comment]
You can call `Arb.register` multiple times with the same type on the same thread; that is a no-op. So as long as you don't run a single test or test suite (whatever your chosen parallelism unit is) on more than one thread, I conservatively call `Arb.register` before each unit (typically test or test suite) and restore it after. That is exactly what passing a number of arbitrary types to `Config` does too.

**Comment by AnthonyLloyd** [comment]
Ok, so its just to be used on a local thread.

I think that leaves me with telling Expecto users that they can't use `Arb.register` since there's no way of registering across all threads that will be created.

**Comment by kurtschelfthout** [comment]
Yes. FsCheck v2 has no built in capacity to run tests in parallel, but assumes test runners can use multiple threads to run sets of tests. Essentially any setup necessary is up to the test runner to provide. Usually register can go into some pre-test or test suite method. 

This will change in v3 where we'll try to keep the registration of types to arbitraries local to each property and without using mutable global state.

**Comment by AnthonyLloyd** [comment]
Ok thanks
